Prevent collection different than Product Catalog from being filtered

kmanijak opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

This PR does a couple of things:

Major:
- Only Product Collection blocks that inherit query from the template are filterable. All the others ones are not. An important note is that only changes in `plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection.php` file are responsible for that.

Minor / consequences:
- Because of the above change I had to adjust some tests so the filters run with Product Collection that inherits the query from the template so we can see the right results hence some changes in test setups.
- In addition, Product Filters tests were based on HandleBars templates rendered on custom pages. Product Collection cannot inherit query from template on pages hence there was a problem as filters were not even rendered there after the change (not sure if that's intended). I extended the utils in `create-dynamic-content.ts` by @samueljseay  with utils related to rendering content into templates and more specifically Product Catalog template.
- Also, I updated handlebars templates:
    - Updated attributes of Product Collections
    - Moved Product Filters outside of Product Collection as the filters should work by default in that configuration

Closes https://github.com/woocommerce/woocommerce/issues/44391

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

### Current filters

1. Go to Product Catalog
2. Insert Product Filters pattern
3. Insert Product Collection
4. Choose Product Catalog collection
5. Insert another Product Collection
6. Choose On Sale collection
7. Save and go to frontend
8. Play around with the filters
9. Make sure ONLY Product Catalog products are being filtered and On Sale collection stays untouched

### [Experimental - to be tested only by dev team!] Product Filters

1. Go to Product Catalog
2. Insert Product Filters (Beta) blocks:
    - Stock Status
    - Price
    - Attribute
    - Rating
4. Insert Product Collection
5. Choose Product Catalog collection
6. Insert another Product Collection
7. Choose On Sale collection
8. Save and go to frontend
9. Play around with the filters
10. Make sure ONLY Product Catalog products are being filtered and On Sale collection stays untouched

<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box and supplying data. -->

-   [ ] Automatically create a changelog entry from the details below.

<details>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

#### Comment <!-- If the changes in this pull request don't warrant a changelog entry, you can alternatively supply a comment here. Note that comments are only accepted with a significance of "Patch" -->

</details>

kmanijak committed:
Prevent collection different than Produyct Catalog from being filtered

kmanijak committed:
Merge branch 'trunk' into 44391-product-collection-update-logic-for-frontend-filters

kmanijak committed:
Adjust Product Filters tests

kmanijak committed:
Split taxonomies filter into seperate filter per taxonomy

kmanijak committed:
Add changelog

kmanijak committed:
Improve onDeselect function

kmanijak committed:
Merge branch 'trunk' into 44391-product-collection-update-logic-for-frontend-filters

kmanijak committed:
Update Product Filter test handlebars so they all use Product Collection that inherits query from template

kmanijak committed:
Set Product Collection to inherit query from template in PHP unit tests

kmanijak committed:
Remove unnecessary empty space

kmanijak committed:
Add changelog

kmanijak committed:
Update tests

kmanijak committed:
Fix lint

kmanijak committed:
Merge branch 'trunk' into 42049-product-collection-break-taxonomies-filter-into-multiple-menu-items

kmanijak committed:
Update Product Collection attributes in filter E2E handlebars

kmanijak committed:
Merge branch '42049-product-collection-break-taxonomies-filter-into-multiple-menu-items' into 44391-product-collection-update-logic-for-frontend-filters

kmanijak committed:
Update the Product Filters testing pages title

kmanijak committed:
Satisfy TS by returning single React element rather than array of elements

kmanijak committed:
Revert some incorrect changes

kmanijak committed:
Cleanup after accidental incorrect merge

kmanijak committed:
Merge branch 'trunk' into 44391-product-collection-update-logic-for-frontend-filters

kmanijak committed:
Add template utils to dynamic contents util

kmanijak committed:
Merge branch 'trunk' into 44391-product-collection-update-logic-for-frontend-filters

kmanijak committed:
Fix Attribute Filter tests

kmanijak committed:
Cleanup Attribute tests, refactor and fix types

kmanijak committed:
Add product filter on top of attirbute filter in handlebars

kmanijak committed:
Fix lint

kmanijak committed:
Update Rating filter tests to use Product Archive template instead of post

kmanijak committed:
Update Rating Filter handlebars

kmanijak committed:
Update Stock filter tests to use Product Archive template instead of post

kmanijak committed:
Update Stock Filter handlebars

kmanijak committed:
Add more granular util to update Produyct Catalog based on handlebar template

kmanijak committed:
Use Product Catalog util and simplify test scenarios

kmanijak committed:
Unify handlebar templates

kmanijak committed:
Fix Product Collection attributes in Product Filter handlebar templates

kmanijak committed:
Revert temporary timeouts increase

kmanijak committed:
Remove unused types

kmanijak committed:
Cleanup types

kmanijak committed:
Cleanup types

kmanijak committed:
Lint fix

thealexandrelara reviewed (approved):
Code is looking good and everything is working as expected.

✅ When the Filters (classic) are changed, only the Product Catalog collection is updated.
✅ When the Product Filters (Beta) block is changed, only the Product Catalog collection is updated.

Considering that, I'm approving the PR. I have some points from my testing that I believe it's not related to this PR, but I'm sharing below:

--- 
I noticed that products are being displayed in the Best Selling and Top Rated collections even though my store didn't sell anything or have any reviews. Is this an expected behavior?

![CleanShot 2024-04-10 at 17 43 59@2x](https://github.com/woocommerce/woocommerce/assets/20469356/dc1f4fa5-01aa-4de7-a90a-331289b6b1b1)

I also noticed that the New Arrivals collection appears empty when there are no recently added products in the store.  Currently, only the collection title shows.  Would it be better to add a message explaining this, or to hide the entire collection when there are no products?

| Editor | Frontend |
|--------|--------|
| ![CleanShot 2024-04-10 at 17 55 41@2x](https://github.com/woocommerce/woocommerce/assets/20469356/14429478-0c78-4f8f-8c2a-abcf6cb4334e) | ![CleanShot 2024-04-10 at 17 50 30@2x](https://github.com/woocommerce/woocommerce/assets/20469356/6bc651cd-d454-4db0-9478-160765d1f59a) |

samueljseay commented on plugins/woocommerce-blocks/tests/e2e/utils/create-dynamic-content.ts:
```diff
@@ -37,8 +29,6 @@ export const deletePost = async ( requestUtils: RequestUtils, id: number ) => {
 
-const posts: number[] = [];
-
 const createPost = async (
```
Hmm, should we really delete this altogether? It seems premature to me. In cases where you don't need template support won't this still be a quick/easy way to create a testing harness for your block?

kmanijak replied:
That is only a cleanup:
- `posts` array is never used except for in line 53 where post IDs are pushed to but never read from.
- `type TestingPost` wasn't used anywhere as well.

Is there some purpose for these vars/types that I missed?

samueljseay replied:
@kmanijak sorry I don't know how I misread the diff on this one. Looks good.

samueljseay commented on plugins/woocommerce-blocks/tests/e2e/utils/create-dynamic-content.ts:
```diff
@@ -74,3 +63,42 @@ export const createPostFromTemplate = async (
 	return createPost( requestUtils, payload );
 };
+
+export const updateTemplatesContent = async (
```
This naming is super confusing to me. Isn't this just creating a template? `createTemplate`?

kmanijak replied:
So this one doesn't create a new template, it actually updates the content of existing template.

Initially, I took the approach of creating custom templates based on [`requestUtils.createTemplate` ](https://github.com/WordPress/gutenberg/blob/e1842bb8dadd224ccf2e7499bcff781afb9ce499/packages/e2e-test-utils-playwright/src/request-utils/index.ts#L177) but then it's impractical to use custom templates in the tests so I gave up on that solution and pivoted towards updating existing (commonly used) templates.

So this util uses the POST method which updates the template content ([documentation](https://developer.wordpress.org/rest-api/reference/wp_templates/#update-a-template)). Hence I think it should be named `update` and not `create`. WDYT?

samueljseay replied:
Damn sorry, I saw the `POST` and assumed it was create, missed that you're passing an ID here.

samueljseay replied:
Nitpick, but I think in this case I think it should be `updateTemplateContents` not `updateTemplatesContent`.

samueljseay commented on plugins/woocommerce-blocks/tests/e2e/utils/create-dynamic-content.ts:
```diff
@@ -74,3 +63,42 @@ export const createPostFromTemplate = async (
+	} );
+};
+
+export const updateProductCatalogContent = async (
```
This should not be here. You should put it with your test and just call `createTemplate` there. This is too specific to the product catalog.

The idea of these helpers is to be generically useful to all e2e tests.

kmanijak replied:
I see where you coming from and that was my initial approach. [Here's the example ](https://github.com/woocommerce/woocommerce/blob/af4bbb9a0ae5ec913dbe082ac380a9e96ac11882/plugins/woocommerce-blocks/tests/e2e/tests/filter-blocks/attribute-filter.block_theme.spec.ts)of Attribute Filter tests at earlier stage of this PR.

It was exactly as you described but then I saw there's a common pattern and code repetition that I could lift to the util. In addition, I think there's a lot of scenarios based on Product Catalog template (the main template in the whole shop) and I agree - it's not generic but at the same time it isn't random one-time use template, hence I thought it may be ok to provide that more specific helper in here.

I'm ok with moving that helper to `tests/filter-blocks/utils.ts`. But at the same time, as we move more and more towards using `create-dynamic-content.ts` I believe there will be more scenarios based on Product Catalog template.

Let me know what you think about it?

samueljseay replied:
@kmanijak I can totally see where you're coming from.

Looking at this with fresh eyes and based on what you're saying I mostly agree. (Apologies I think I didn't spend enough time to understand the approach originally).

So now I better understand what you're doing I think the only thing that bothers me is the hard-coded `/shop` link. I wonder if we really need to `use` anything in this situation. Since a template and a page are sort of decoupled from each other, maybe we can just hardcode navigating to the page in the test and remove the API where we give a `link` via the test fixture? 

Additionally, would it be unreasonable to just call this `updateProductCatalogTemplate`?

kmanijak replied:
> So now I better understand what you're doing I think the only thing that bothers me is the hard-coded /shop link. I wonder if we really need to use anything in this situation. Since a template and a page are sort of decoupled from each other, maybe we can just hardcode navigating to the page in the test and remove the API where we give a link via the test fixture?

We could remove the link from `util` and hardcode it in the test but I still need to import the particular template instance in a test so the test uses a correct template version.

<img width="784" alt="image" src="https://github.com/woocommerce/woocommerce/assets/20098064/fa89e25b-b243-4420-981b-01abf344db5c">

For example in `attribute-filter.block_theme.spec.ts` we have three template specs: `defaultBlockTemplate`, `templateWithShowCounts`, `dropdownBlockTemplate`. With the template, the issue is they use the same instance hence I didn't find an alternative way to let the test know which "version" of the template it should use. With post it was easy, you use specific (and unique) post instances.

samueljseay replied:
@kmanijak I think thats why it doesn't make much sense to use fixtures here? You can set up the template once in the test? or if multiple tests use a template you can group them with decribe and use beforeAll?

samueljseay commented on plugins/woocommerce-blocks/tests/e2e/playwright-utils/test.ts:
```diff
@@ -210,9 +222,29 @@ const test = base.extend<
+					data
+				);
+
+			const utilUpdateProductCatalogContent = (
```
Exposing this here is not the spirit of a generically useful testing util. Just expose the createTemplate stuff and add your test specific content creation in your test instead.

samueljseay commented on plugins/woocommerce-blocks/tests/e2e/playwright-utils/test.ts:
```diff
@@ -210,9 +222,29 @@ const test = base.extend<
 			const utilDeletePost = ( id: number ) =>
 				deletePost( requestUtils, id );
 
+			const utilUpdateTemplatesContent = (
```
Again naming here. This should surely just be createTemplate.

samueljseay reviewed (changes requested):
Just some thoughts from me on the test helpers.

kmanijak commented:
@thealexandrelara thanks for the review! Regarding the points you raised:

> I noticed that products are being displayed in the Best Selling and Top Rated collections even though my store didn't sell anything or have any reviews. Is this an expected behavior?

🤔 I don't think I have a good answer to that.

But happy to get feedback from @manospsyx or @jarekmorawski: do you folks think Top Rated or Best Selling collections should warrant products are actually rated or have at least one sell?

My assumption is it's a bit of an edge case for the early stages of the stores and maybe it's ok to display _some_ products until there are first ratings/first sells.

> I also noticed that the New Arrivals collection appears empty when there are no recently added products in the store. Currently, only the collection title shows. Would it be better to add a message explaining this, or to hide the entire collection when there are no products?

Right. This is because the New Arrivals collection displays products from the last 7 days by default. Users can change it if needed.

<img width="283" alt="image" src="https://github.com/woocommerce/woocommerce/assets/20098064/938b2462-ca38-45fa-adf1-5ea704ba520b">

But I agree the experience can be improved and we're aiming to implement your latter suggestion with https://github.com/woocommerce/woocommerce/issues/43673.

jarekmorawski commented:
> do you folks think Top Rated or Best Selling collections should warrant products are actually rated or have at least one sell?

Good question and a very valid one given that the block can be used by stores at different stages of growth. I think we should ask the user what they want to do when no products meet the criteria. Here's a quick mockup.

![pg2](https://github.com/woocommerce/woocommerce/assets/79307566/da7ccf1f-2f76-4d66-a830-02b80744e672)

manospsyx commented:
Let's avoid offering any random recommendations / results in the frontend, other than the ones that precisely match the intended criteria set at the Collection level by the merchant :) 

In the Editor, we'll want to show placeholder/random products to allow the user to make customizations, until the collection is able to generate the intended recommendations. At the same time, we can indicate that results are "previews/placeholders" of the actual results, and explain that:

- no results were found matching the specified criteria (perhaps a different preview label tooltip/state?);
- perhaps encourage the merchant to use the "No results" block to define fallback recommendations until some products match the specified criteria (e.g. merchants could add a different collection there as a fallback).

kmanijak committed:
Merge branch 'trunk' into 44391-product-collection-update-logic-for-frontend-filters

kmanijak committed:
Rename updateTemplatesContent to updateTemplateContents to better depict its purpose

kmanijak committed:
Fix types

kmanijak committed:
Rename updateProductCatalogContent to updateProductCatalogTemplate to better depict its purpose

kmanijak committed:
Merge branch 'trunk' into 44391-product-collection-update-logic-for-frontend-filters

kmanijak committed:
Merge branch 'trunk' into 44391-product-collection-update-logic-for-frontend-filters

kmanijak committed:
Allow for beforeAll and afterAll in PW tests

kmanijak committed:
Use beforeAll and afterAll instead of using fixture in active filters test

kmanijak committed:
Use beforeAll and afterAll instead of using fixture in attribute filters test

kmanijak committed:
Use beforeAll and afterAll instead of using fixture in price filters test

kmanijak committed:
Use beforeAll and afterAll instead of using fixture in rating filters test

kmanijak committed:
Use beforeAll and afterAll instead of using fixture in stock status filters test

kmanijak committed:
Get rid of updateProductCatalogTemplate util as it's too specific

kmanijak committed:
Extract some constants

kmanijak committed:
Remove unused file probably added here by accident

kmanijak committed:
Adjust Attribute Filter block tests to the new logic

kmanijak committed:
Merge branch 'trunk' into 44391-product-collection-update-logic-for-frontend-filters

kmanijak committed:
Merge fix

kmanijak committed:
Bring back necessary handlebars file

kmanijak committed:
Merge fix

kmanijak committed:
Merge fix

kmanijak committed:
Update beforeAll

kmanijak committed:
Avoid relying on product collection page object in before all

kmanijak committed:
Bring back the original order of tests

kmanijak committed:
Switch Filter by Attribute tests to handlebars

kmanijak committed:
Fix typo

kmanijak committed:
Update test to new circumstances

WunderBart committed:
Merge remote-tracking branch 'origin' into 44391-product-collection-update-logic-for-frontend-filters

WunderBart committed:
Fix hooks usage
We no longer use hooks other than `beforeEach` since we're resetting the DB for each test (#46125). Using other hooks was not caught by the linter (on this PR) because I accidentally removed the no-hooks rule in #47228 and restored it just now in #47500. 😅

kmanijak committed:
Fix the test

kmanijak committed:
Fix Price filter test

kmanijak committed:
Update attribute filter tests

kmanijak committed:
Share the handlebars between tests

kmanijak committed:
Update price filter tests

kmanijak committed:
Update rating filter tests

kmanijak committed:
Update stock filter tests

kmanijak committed:
Merge branch 'trunk' into 44391-product-collection-update-logic-for-frontend-filters

kmanijak committed:
Remove unnecessary product collection page object from filter tests

kmanijak committed:
Fix typo

kmanijak committed:
Fix Filter by Stock test

kmanijak committed:
Simplify Attribute filter test by remoiving one filter from page

kmanijak committed:
Try to wait for element to be visible

kmanijak committed:
Fix handlebars syntax for filters so it doesnt throw validation errors and warnings

kmanijak committed:
Wait for filters to be visible for sure before performing further part of test

WunderBart committed:
Wait for Site Editor canvas loader

WunderBart committed:
Fix linter error

WunderBart committed:
Merge remote-tracking branch 'origin' into 44391-product-collection-update-logic-for-frontend-filters

kmanijak commented:
> @kmanijak I think thats why it doesn't make much sense to use fixtures here? You can set up the template once in the test? or if multiple tests use a template you can group them with decribe and use beforeAll?

@samueljseay, it's been a while but I addressed your feedback and removed the fixtures as well as Product Catalog-specific utils. There's only a single generic util for updating template content used across many within `beforeEach` (it's expected to be in `beforeEach` and not `beforeAll` after changes introduced by @WunderBart, where DB is reset before each test).

Can I ask for re-review?

Also, either @imanish003 or @thealexandrelara , do you mind reviewing it? I'd love to merge it before end of week 🙏

WunderBart reviewed (approved):
🚀

imanish003 reviewed (approved):
I tested the changes with old filters as well as new filters. Everything seems to be working as expected. Well done! 🚀

samueljseay reviewed (approved):
Looking good from my perspective. Thanks for addressing all my concerns 🚢
